### ISSUE TITLE
pfc.ko build isn't re-attempted if it fails (but libpfc.so succeeds) when re-running make

### ISSUE THREAD

**Issue by pcordes** [travisdowns/uarch-bench#24]
I first tried `make -j8` and didn't notice right away that there were errors.  libpfc failed to build, so make stopped before building uarch-bench.  That's fine, but it doesn't retry building libpfc.

The cause of libpfc failing to build might be a libpfc bug, though.

```
$ make -j8
...
cc  -g -Wall -Werror -Wextra -Wno-unused-parameter -I. -I/usr/local/src/uarch-bench/libpfm-4.8.0/lib/../include -DCONFIG_PFMLIB_DEBUG -DCONFIG_PFMLIB_OS_LINUX -D_REENTRANT -I. -fvisibility=hidden -DCONFIG_PFMLIB_ARCH_X86 -DCONFIG_PFMLIB_ARCH_X86_64 -I. -c pfmlib_intel_snb.c
rm -rf kmod.build
cp -r kmod kmod.build
cd kmod.build && make MAKEFLAGS=
make[2]: Entering directory '/usr/local/src/uarch-bench/libpfc/kmod.build'
make -C /lib/modules/4.12.8-2-ARCH/build M=/usr/local/src/uarch-bench/libpfc/kmod.build modules
make[3]: Entering directory '/usr/lib/modules/4.12.8-2-ARCH/build'
make[3]: *** No rule to make target 'modules'.  Stop.
make[3]: Leaving directory '/usr/lib/modules/4.12.8-2-ARCH/build'
make[2]: *** [Makefile:5: all] Error 2
make[2]: Leaving directory '/usr/local/src/uarch-bench/libpfc/kmod.build'
make[1]: *** [Makefile:30: pfc.ko] Error 2
make[1]: Leaving directory '/usr/local/src/uarch-bench/libpfc'
make: *** [Makefile:66: libpfc/libpfc.so] Error 2
cc  -g -Wall -Werror -Wextra -Wno-unused-parameter -I. -I/usr/local/src/uarch-bench/libpfm-4.8.0/lib/../include -DCONFIG_PFMLIB_DEBUG -DCONFIG_PFMLIB_OS_LINUX -D_REENTRANT -I. -fvisibility=hidden -DCONFIG_PFMLIB_ARCH_X86 -DCONFIG_PFMLIB_ARCH_X86_64 -I. -c pfmlib_intel_snb_unc.c
make: *** Waiting for unfinished jobs....
cc  -g -Wall -Werror -Wextra -Wno-unused-parameter -I. -I/usr/local/src/uarch-bench/libpfm-4.8.0/lib/../include -DCONFIG_PFMLIB_DEBUG -DCONFIG_PFMLIB_OS_LINUX -D_REENTRANT -I. -fvisibility=hidden -DCONFIG_PFMLIB_ARCH_X86 -DCONFIG_PFMLIB_ARCH_X86_64 -I. -c pfmlib_intel_ivb.c
cc  -g -Wall -Werror -Wextra -Wno-unused-parameter -I. -I/usr/local/src/uarch-bench/libpfm-4.8.0/lib/../include -DCONFIG_PFMLIB_DEBUG -DCONFIG_PFMLIB_OS_LINUX -D_REENTRANT -I. -fvisibility=hidden -DCONFIG_PFMLIB_ARCH_X86 -DCONFIG_PFMLIB_ARCH_X86_64 -I. -c pfmlib_intel_ivb_unc.c
cc  -g -Wall -Werror -Wextra -Wno-unused-parameter -I. -I/usr/local/src/uarch-bench/libpfm-4.8.0/lib/../include -DCONFIG_PFMLIB_DEBUG -DCONFIG_PFMLIB_OS_LI
```

Running `make` again after this built the `uarch-bench` binary.

But pfc kmod isn't built, so I guess somewhere there's a missing dependency to check on that?

Also I'm having trouble getting libpfc built on Arch Linux.  I `cd`ed into `uarch-bench/libpfc` and did a `make clean`, then `make`:

```
rm -rf kmod.build
cp -r kmod kmod.build
cd kmod.build && make MAKEFLAGS=
make[1]: Entering directory '/usr/local/src/uarch-bench/libpfc/kmod.build'
make -C /lib/modules/4.12.8-2-ARCH/build M=/usr/local/src/uarch-bench/libpfc/kmod.build modules
make[2]: Entering directory '/usr/lib/modules/4.12.8-2-ARCH/build'
make[2]: *** No rule to make target 'modules'.  Stop.
make[2]: Leaving directory '/usr/lib/modules/4.12.8-2-ARCH/build'
make[1]: *** [Makefile:5: all] Error 2
make[1]: Leaving directory '/usr/local/src/uarch-bench/libpfc/kmod.build'
make: *** [Makefile:30: pfc.ko] Error 2
```

I don't see how the `make -C /lib/modules/$(shell uname -r)/build M=$(PWD) modules` is supposed to work, because there's no `modules` anywhere in libpfc or in `/usr/lib/modules/4.12.8-2-ARCH/`, just `modules.*`.  Same on my Ubuntu 15.04 machine using a distro kernel.  (I keep putting off updating it until I make a grand plan to put it on an SSD or something...)

And the `Makefile` doesn't define this target.

**Comment by travisdowns** [comment]
I'll split this up into two issues, for the "not rebuilding" and the "libpfc doesn't build" issue.

**Comment by travisdowns** [comment]
There is no perfect, simple, system to build submodules in make. In the makefile the dependency on libpfc is expressed via the libpfc shared library: `libpfc/libpfc.so`. The project has other outputs, like `pfc.ko` and if for some reason `libpfc.so` succeeds and `pfc.ko` fails, it won't retry as you've noticed.

For now use either `make dist-clean && make` or just run make directly in the `libpfc` subdir as you've been doing.

**Comment by travisdowns** [comment]
@pcordes - I created the bug above to get the kernel module working, but I forgot to mention that if you just want to run the test, you can skip the `libpfc` stuff entirely and just use the clock-based timer: it calibrates to cycles based on a timing loop and is very also very accurate as long as frequency scaling is disabled.

To use the clock timer, just omit the `--timer` argument and run it like: `./uarch-bench.sh --test-name=add-loop`. I get the following results:

```
Welcome to uarch-bench (3cc7031-dirty)
Median CPU speed: 3.492 GHz
Running 12 benchmark groups using timer clock

** Running benchmark group default **
                     Benchmark   Cycles    Nanos

** Running benchmark group misc **
                     Benchmark   Cycles    Nanos
               32-bit add-loop     1.00     0.29
               64-bit add-loop     1.00     0.29
```

... and I didn't even turn off turboboost. 

You can disable `libpfc` in the build by changing the `USE_LIBPFC` line to `0` in `config.mk`.

**Comment by travisdowns** [comment]
The kind of ugly hack we used here is to add `pfc.ko` as a dependency of `uarch-bench` when `USE_LIBPFC == 0`, in the same way we do for `libpfc.so`. That part isn't ugly. The ugly part is that this means that in concurrent make, it will try to build both `pfc.ko` and `libpfc.so` in parallel using the same rule, which will blow up (since it results in building the `libpfc` project twice at the same time, so everything stomps on everything else).

The hack to fix that is to make `pfc.ko` depend on `libpfc.so` in the uarch-bench `Makefile`, so that the libpfc make doesn't happen in parallel. The second make usually still happens (because the `pfc.ko` timestamp is later than the `libpfc.so` one), but it does nothing.
